subprojects/core/src/main/java/org/gradle/api/internal/collections/IterationOrderRetainingSetElementSource.java

```diff
@@ -16,75 +16,138 @@
 
 package org.gradle.api.internal.collections;
 
+import com.google.common.base.Objects;
 import org.gradle.api.Action;
+import org.gradle.api.internal.provider.AbstractProvider;
 import org.gradle.api.internal.provider.ProviderInternal;
 
+import javax.annotation.Nullable;
 import java.util.Collection;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedHashSet;
 import java.util.Set;
 
 public class IterationOrderRetainingSetElementSource<T> implements ElementSource<T> {
-    private final Set<T> values = new LinkedHashSet<T>();
+    // This set represents the elements which have been realized and "added" to the store.  This may
+    // or may not be in the order that elements are inserted.  We track this to know which realized
+    // elements have already had rules fired against them.
+    private final Set<T> added = new HashSet<T>();
+
+    // This set represents the order in which elements are inserted to the store, either actual
+    // or provided.  We construct a correct iteration order from this set.
+    private final Set<RealizableProvider<T>> inserted = new LinkedHashSet<RealizableProvider<T>>();
```

Comment:
As far as I can see, both this and "orderedValuesCache" could be a List to save a lot of memory. The set semantics seem to be guarded by the other sets already. Did I miss something?